src/QsCompiler/Transformations/CallGraph/CallGraphWalker.cs

```diff
@@ -0,0 +1,507 @@
+                    var typeRes = combination.CombinedResolutionDictionary.FilterByOrigin(identifier);
+                    this.ExpTypeParamResolutions = new List<TypeParameterResolutions>();
+
+                    var referenceRange = Range.Zero;
+                    if (this.CurrentStatementOffset.IsValue
+                        && this.CurrentExpressionRange.IsValue)
+                    {
+                        referenceRange = this.CurrentStatementOffset.Item + this.CurrentExpressionRange.Item;
+                    }
+
+                    var called = new SimpleCallGraphNode(identifier);
+                    var edge = new SimpleCallGraphEdge(typeRes, this.CurrentNode.CallableName, called.CallableName, referenceRange);
+                    this.Graph.AddDependency(this.CurrentNode, called, edge);
+                    // If we are not processing all elements, then we need to keep track of what elements
+                    // have been processed, and which elements still need to be processed.
+                    if (this.WithTrimming
+                        && !this.RequestStack.Contains(called)
+                        && !this.ResolvedNodeSet.Contains(called))
+                    {
+                        this.RequestStack.Push(called);
+                    }
+                }
+            }
+
+            private class NamespaceWalker : NamespaceTransformation<TransformationState>
+            {
+                public NamespaceWalker(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override QsCallable OnCallableDeclaration(QsCallable c)
+                {
+                    if (!this.SharedState.WithTrimming)
+                    {
+                        var node = new SimpleCallGraphNode(c.FullName);
+                        this.SharedState.CurrentNode = node;
+                        this.SharedState.Graph.AddNode(node);
+                    }
+                    return base.OnCallableDeclaration(c);
+                }
+            }
+
+            private class ExpressionKindWalker : ExpressionKindTransformation<TransformationState>
+            {
+                public ExpressionKindWalker(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override ExpressionKind OnIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
+                {
+                    if (sym is Identifier.GlobalCallable global)
+                    {
+                        this.SharedState.AddDependency(global.Item);
+                    }
+
+                    return ExpressionKind.InvalidExpr;
+                }
+            }
+        }
+
+        private static class ConcreteCallGraphWalker
```

Comment:
Similar to `SimpleCallGraph` and `ConcreteCallGraph`, it might be cleaner to have the simple and concrete walkers in separate files as well.

Reply:
Here they are private members of a static class. I don't want to have to expose them.

Reply:
They would only be exposed internally at most. The reason I'm asking is because there are 15 classes in this file at varying levels of nesting which makes it a little hard to follow the logic.

Reply:
done